libraries/AP_OpticalFlow/AP_OpticalFlow_UPFLOW.cpp

```diff
@@ -87,6 +259,91 @@ void AP_OpticalFlow_UPFLOW::init()
     }
     // open serial port with baud rate of 19200
     uart->begin(19200);
+
+    /*
+        LC306 or LC302-8B boards require initialization before use, but LC302-3C does not.
+        Initialization can divided into four parts
+
+        (0xAA)start --> (0xAB)internal parameter configuration
+        --> (0xBB)sensor parameter configuration (Loop) --> (0xDD)End
+            ^--------------------------------------/
+
+        | command sent | returned status |
+        | ------------ | --------------- |
+        |     0xAA     |      none       | 1. Start flag
+        | ------------ | --------------- |
+        |     0xAB     |      0xAB       | 2. internal parameter configuration flag
+        |     0x96     |                 | status code:  0x00 == success
+        |     0x26     |   status code   |               0x01 == IIC no response
+        |     0xbc     |                 |               0x02 == XOR error
+        |     0x50     |                 | send: 0xAB, 0x96, 0x26, 0xbc, 0x50, 0x5c(BCC without 0xAB)
+        |   BCC(0x5c)  | 0xAB xor status | success response: 0xAB, 0x00, 0xAB
+        | ------------ | --------------- |
+        |     0xBB     |      0xBB       | 3. sensor parameter configuration
+        |     0xdc     |                 |    sending data in upflow_sensor_cfg
+        |    address   |   status code   | same status code above
+        |     data     |                 | send: 0xBB, 0xdc, 0xXX, 0xXX, 0xXX(BCC without 0xBB)
+        |     BCC      | 0xBB xor status | success response: 0xBB, 0x00, 0xBB
+        | ------------ | --------------- |
+        |     0xDD     |      none       | 4. Stop flag
+
+        !:  Step 3 will run repeatedly until the sending is complete.
+            Each time the module receives an 0xBB instruction, it will return a corresponding status.
+    */
+
+    // uint32_t nbytes = 0;
+    uint8_t bcc_sum = 0;
+    uint8_t recv_buf[3];
+
+    //delay 100ms before initialization
+    hal.scheduler->delay(100);
+
+    // GCS_SEND_TEXT(MAV_SEVERITY_INFO, "UPFLOW: Start initialization!");
+
+    // (0xAA)start configuration
+    uart->write((uint8_t)0xAA);
+
+    // (0xAB)internal parameter configuration
+    uart->write((uint8_t)0xAB);
+    for (int i = 0; i < 4; i++) {
+        uart->write(upflow_internal_para[i]);
+        bcc_sum ^= upflow_internal_para[i];
+    }
+    uart->write(bcc_sum);
+    hal.scheduler->delay(10);
+    for (int i = 0; i < 3; i++) {
+        recv_buf[i] = uart->read();
+    }
+    // if failed to receive response code, you can start over from 0xAA.
+    if ((recv_buf[0] ^ recv_buf[1]) != recv_buf[2]) {
+        UPFLOW_INIT_FAILED;
+        return;
```

Comment:
What happens on this early return? We don't call init a second time, so does this mean the sensor never works ?

Reply:
> What happens on this early return? We don't call init a second time, so does this mean the sensor never works ?

The reason for early return is that if the user inserts the LC302-3C, the sensor will not return any data during the initialization operation, and there is no need to continue the initialization operation. LC302-3C and other sensor versions that do not require initialization will not be affected. 

The _UPFLOW_INIT_FAILED_ macro is what I've set during debugging to monitor whether the initialization is successful or not in the ground station.
And I have conducted multiple tests using the Pixhawk 2.4.8 flight controller. For LC306 and LC302-8B, they can be initialized without any issues.